src/controller/domain_handle.rs

```diff
@@ -59,7 +59,8 @@ impl<'a> BatchSendHandle<'a> {
             }
             let key_col = key[0];
 
-            let mut shard_writes = vec![Vec::new(); self.dih.txs.len()];
+            let mut shard_writes: HashMap<usize, Vec<Record>> =
+                HashMap::with_capacity(self.dih.txs.len());
```

Comment:
I think I'd prefer for this to say a `Vec<Vec<_>>`. Any particular reason you want it to be a `HashMap`? I guess if there are *a lot* of shards?